File: SlapMac-Android/app/src/main/kotlin/com/slapmac/app/SlapDetector.kt
package com.slapmac.app

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import kotlin.math.abs
import kotlin.math.sqrt

class SlapDetector(context: Context) : SensorEventListener {
    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)

    var sensitivity = 1.5
    var cooldownMs = 1500L
    var onSlapDetected: (() -> Unit)? = null
    var isRunning = false
        private set

    private var lastSlapTime = 0L
    private var previousMagnitude = 0.0
    private var baselineMagnitude = 9.81
    private var calibrationCount = 0
    private val calibrationTotal = 50

    fun start() {
        if (accelerometer == null || isRunning) return
        isRunning = true
        calibrationCount = 0
        baselineMagnitude = 9.81
        sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_GAME)
    }

    fun stop() {
        isRunning = false
        sensorManager.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        if (!isRunning || event.sensor.type != Sensor.TYPE_ACCELEROMETER) return

        val x = event.values[0].toDouble()
        val y = event.values[1].toDouble()
        val z = event.values[2].toDouble()
        val magnitude = sqrt(x * x + y * y + z * z)

        // Calibration phase
        if (calibrationCount < calibrationTotal) {
            baselineMagnitude = maxOf(baselineMagnitude, magnitude)
            calibrationCount++
            previousMagnitude = magnitude
            return
        }

        val delta = abs(magnitude - previousMagnitude)
        previousMagnitude = magnitude

        // Adapt baseline slowly
        baselineMagnitude = baselineMagnitude * 0.98 + magnitude * 0.02

        // Detection threshold inversely proportional to sensitivity
        val threshold = 2.0 / sensitivity.coerceIn(0.5, 4.0)

        if (delta > threshold) {
            val now = System.currentTimeMillis()
            // Extended suppression: 1 slap = 1 sound
            val suppressionMs = maxOf(cooldownMs * 2, 3000L)
            if (now - lastSlapTime >= suppressionMs) {
                lastSlapTime = now
                onSlapDetected?.invoke()
            }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
}

File: SlapMac-iOS/SlapMac/SlapDetector.swift
import Foundation
import CoreMotion

final class SlapDetector: ObservableObject {
    private let motionManager = CMMotionManager()
    private let processingQueue = OperationQueue()

    private var lastSlapTime = Date.distantPast
    private var previousMagnitude: Double = 0
    private var baselineMagnitude: Double = 1.0
    private var calibrationCount = 0
    private let calibrationTotal = 50

    @Published var isRunning = false

    var sensitivity: Double = 1.5
    var cooldownInterval: TimeInterval = 1.5
    var onSlapDetected: (() -> Void)?

    var isAvailable: Bool {
        motionManager.isAccelerometerAvailable
    }

    init() {
        processingQueue.name = "com.slapmac.motion"
        processingQueue.maxConcurrentOperationCount = 1
    }

    func start() {
        guard motionManager.isAccelerometerAvailable, !isRunning else { return }

        DispatchQueue.main.async { self.isRunning = true }
        calibrationCount = 0
        baselineMagnitude = 1.0

        motionManager.accelerometerUpdateInterval = 0.01
        motionManager.startAccelerometerUpdates(to: processingQueue) { [weak self] data, _ in
            guard let self, let data else { return }
            self.processAcceleration(data.acceleration)
        }
    }

    func stop() {
        motionManager.stopAccelerometerUpdates()
        DispatchQueue.main.async { self.isRunning = false }
    }

    private func processAcceleration(_ accel: CMAcceleration) {
        let magnitude = sqrt(accel.x * accel.x + accel.y * accel.y + accel.z * accel.z)

        // Calibration phase
        if calibrationCount < calibrationTotal {
            baselineMagnitude = max(baselineMagnitude, magnitude)
            calibrationCount += 1
            previousMagnitude = magnitude
            return
        }

        let delta = abs(magnitude - previousMagnitude)
        previousMagnitude = magnitude

        // Slowly adapt baseline to ambient motion
        baselineMagnitude = baselineMagnitude * 0.98 + magnitude * 0.02

        // Detection threshold inversely proportional to sensitivity
        let threshold = 2.0 / max(0.5, min(4.0, sensitivity))

        if delta > threshold {
            let now = Date()
            // Extended suppression to guarantee 1 slap = 1 sound
            let suppressionInterval = max(cooldownInterval * 2, 3.0)
            guard now.timeIntervalSince(lastSlapTime) >= suppressionInterval else { return }
            lastSlapTime = now

            DispatchQueue.main.async {
                self.onSlapDetected?()
            }
        }
    }
}
